image-registry/automate-pruning.md

```diff
@@ -0,0 +1,266 @@
+
+#### Orphaned blobs
+
+As an OpenShift cluster admin
+I want the cluster to automatically remove orphaned blobs stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies the current image pruning process invoked by the
+`oc adm prune` command. Running this in a container can be accomplished by
+using the `cli` imagestream installed by the samples operator. The image must
+be run using a service account with the `system:image-pruner` cluster role.
+
+The image registry operator will be updated to configure and reconcile all the
+components needed to run image pruning on an OpenShift cluster.
+Specifically, this includes the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
+   1. Schedule for `CronJob`
+   2. `--keep-tag-revision` command line argument for the pruner
+   3. `--keep-younger-than` command line argument for the pruner
+   4. Resources for the `CronJob` pod template
+
+These components should be created/reconciled when the image registry operator
+is in the `Managed` state. When the image registry operator is in the `Removed`
+state, the above components should be deleted from the cluster if present.
+
+### Risks and Mitigations
+
+#### Resource consumption
+
+Risk: Image pruning requires siginficant memory to compute the blob layer graph
+needed to identify "prunable" layers.
+
+Mitigation: API will allow cluster admins to adjust the resources allocated to
+the `CronJob`.
+
+Risk: Image pruning requires too much memory to be scheduled to a node
+Risk: Image pruning exhausts resources used for application workloads
+
+Mitigation: API will allow admins to use standard tools to let pruning be run
+on desired machine pools (`affinity`, `tolerations`, `nodeSelector`)
+
+##### Concurrent pruning jobs
+
+Risk: Pruning jobs cannot run concurrently.
+Risk: Pruning job durations can exceed the interval set by the `CronJob`
+schedule.
+
+Mitigation: The `CronJob` for pruning must be configured with
+`concurrencyPolicy: Forbid`
+
+#### Removal of past pruning jobs
+
+Risk: Runs of the pruning job create another set of objects in etcd that need
+to be pruned.
+
+Mitigation: `CronJob` specifies a history limit for successful and failed jobs.
+This will be a tunable parameter via the API.
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+Proposed changes to the image registry operator's config spec:
+
+```yaml
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
```

Comment:
@bparees  @openshift/openshift-team-developer-experience  I've struggled with this part of the API.

We initially discussed having the `schedule` field be what determines if the CronJob is enabled or suspended. However it may be more prudent to have a separate field `enabled` that is the primary flag to set. `schedule` could then default to "daily at midnight" if `enabled: true` and schedule is left empty.

Reply:
What about adding a ManagementState? It actually can be "Managed", "Unmanaged" and "Removed".

Reply:
@adambkaplan @bparees it's the other way around. I envision the pruner should have its own configuration resource and the operator should manage them independently. The pruner is not part of the registry, it's more about the image API and if someone decides to remove the registry from their cluster, they still may need to have pruner. For example, an image stream may become big because of scheduled imports. So I like the idea of having a separate operator for pruning (and we already have the repository cluster-image-pruner-operator).

Reply:
at first blush I agree with @dmage here .... mixing more than one config obj (this new one, and the image registry operator's existing one) under the auspices of a single operator seems messy

If the pruner needs to some how interface with the image registry operator and what it manages, those specifics should be called out / understood now and then exposed in the image registry operator's config object if need be

Reply:
While I can see the value in that distinction, please keep in mind the implementation cost of maintaining yet another operator, image, repo, CI jobs, ART process, etc.

Do we think a lot of users are going to use imagestreams but not the registry?  Can we live with telling that subset of users they are on their own to do pruning?  The cases where people blow up their cluster w/ images, as far as i know, are cases where they are using our registry and keep building+pushing new images to it.

Alternatively you can have a separate CRD for pruning if you want, without creating a separate operator.  Of course then you get into weirdness where the registry CRD says "removed" but they still have a valid pruning configuration that the operator is processing, and the telemeter status for the operator would report "removed" even though in reality it's still doing some stuff, so that feels a bit ugly.

You could of course also implement the actual "operator" logic in the openshift-controller (again, introducing a separate CRD for configuring automated pruning).  There's really no reason it needs its own explicit CVO-managed operator.  The main benefit of having a CVO-managed operator is easy telemeter reporting of whether it's healthy or not, imho.

Reply:
> Do we think a lot of users are going to use imagestreams but not the registry? Can we live with telling that subset of users they are on their own to do pruning? 

@siamaksade any insights here?

Reply:
@openshift/api-reviewers if we declare a field as Tech Preview on a `v1` API object, can we remove it in a subsequent release? Such as:

1. 4.3 add `pruning:` as a tech preview field on `config.imageregistry.operator.openshift.io`
2. 4.4 move pruning API to a new GA CRD (`config.imagepruner.operator.openshift.io`) - deprecate `pruning` object in image registry operator CRD.
3. 4.5 delete `pruning` from image registry operator CRD.

Reply:
@adambkaplan there are customers who use imagestreams without the internal registry but they are not the majority. I'm not aware of more granular data on this. That said, I don't think we should target pruning external registries yet. The asks we currently have are for pruning the internal registry as @bparees stated.

Reply:
the question isn't about pruning external registries.  it's about pruning image metadata we've imported from external registries, because it takes up space in your etcd, even if it's not taking up space in your registry storage.

Reply:
In case of the internal registry, auto-pruning removes also the images as I understand. Is that correct?

Reply:
first let's fix our terminology.

There are two critical pieces of data:

1) image/imagestream resource objects in etcd.  These exist both for images stored in the internal registry, as well as when you import an image from an external registry (which again, just means importing the metadata, not the actual image).

2) image content/blobs stored in the registry storage.  These only exist for images that were pushed to the internal registry, or that the internal registry did a "pullthrough" on.

When you run manual pruning (oc adm prune) you can choose whether you want to clean up (1), (2), or (1+2).

Automated pruning is not, i do not think, going to give you that choice.  It will always clean up (1+2).

In the case of external images, where we only have (1), then only (1) will be cleaned up, of course.

Reply:
> Automated pruning is not, i do not think, going to give you that choice. It will always clean up (1+2).

This was something the team discussed at scrum today. If we separate the pruner operator from the registry operator, we can be more intelligent (though still take away the choice):

1. If registry is `Managed`, 1+2 for items on the internal registry, 1 for external images
2. If registry is `Removed`, 1 only (everything should be an external image).

Bundling the pruner as a component installed by the registry operator makes is harder to build an API that lets us have "1 only".

Reply:
you don't have to separate the operator to accomplish that.

i don't even think you have to separate the api, strictly speaking (no reason you can't set the management state to removed, but then have pruning configured, within the same CR..it's a bit weird, but you get to choose the meaning/implications).

Anyway "removed" refers to the operand.  In the case of this particular CR, the "operand" is the registry.

There is no reason we can't have a single operator with multiple operands (the registry being one, the pruner being the other).

Heck, you can even have multiple ClusterVersion objects being created/managed by your single operator. (which would allow you to report degraded status for pruning independent of degraded status for the registry).

So I suggest you go with a distinct CRD for pruning configuration.  Whether that CRD has the traditional "managementstate" field or not, i'm fairly indifferent but leaning towards not having it.

And whether you want to create an additional ClusterVersion object for reporting the health of the "pruning" operand, i'm also largely indifferent, i'd suggest deferring that for now.

I'd start by just firing alerts when pruning goes wrong.

Reply:
Another reason for having a separate CRD for the pruner - it gives us the flexibility to land this feature as Tech Preview. Per @deads2k we can't have tech preview fields on GA APIs.

I think we've reached consensus that this API should live in a separate CRD - I'll update the enhancement.